File: advanced-programming-methods/toy-interpreter-java/src/model/expression/VarExpression.java
package model.expression;

import exceptions.DictionaryException;
import exceptions.ExpressionException;
import model.IDictionary;
import model.IHeap;

/**
 * The type Var expression.
 */
public class VarExpression extends Expression
{
    private String varName;

    /**
     * Instantiates a new Var expression.
     *
     * @param varName the var name
     */
    public VarExpression(String varName)
    {
        this.varName = varName;
    }

    /** {@inheritDoc} */
    @Override
    public int eval(IDictionary table, IHeap heap) throws ExpressionException
    {
        try
        {
            return table.get(varName);
        }
        catch (DictionaryException e)
        {
            throw new ExpressionException(
                String.format("Variable = '%s' is not defined", varName)
            );
        }
    }

    /** {@inheritDoc} */
    @Override
    public String toString()
    {
        return varName;
    }

    /** {@inheritDoc} */
    @Override
    public Expression cloneDeep()
    {
        return new VarExpression(varName);
    }
}

File: advanced-programming-methods/labs/toy-interpreter-sharp/model/expression/VarExpression.cs
using System;
using Exceptions;

namespace Model
{
    /// <summary>
    /// The type Var expression.
    /// </summary>
    [Serializable()]
    public class VarExpression : Expression
    {
        private readonly string varName;

        /// <summary>
        /// Instantiates a new Var expression.
        /// </summary>
        /// <param name="varName"> the var name </param>
        public VarExpression(string varName)
        {
            this.varName = varName;
        }

        /// <inheritdoc />
        public override int Eval(IDictionary table, IHeap heap)
        {
            try
            {
                return table.Get(varName);
            }
            catch (DictionaryException)
            {
                throw new ExpressionException(string.Format("Variable = '{0}' is not defined", varName));
            }
        }

        /// <inheritdoc />
        public override string ToString()
        {
            return varName;
        }

        /// <inheritdoc />
        public override Expression CloneDeep()
        {
            return new VarExpression(varName);
        }
    }

}

